prelude.dx

```diff
@@ -34,36 +34,78 @@ def IToI8  (x : Int)     : Int8    = internalCast _ x
 def IToF (x:Int) : Float = internalCast _ x
 def FToI (x:Float) : Int = internalCast _ x
 
-data Add a:Type =
-  MkAdd (a->a->a) (a->a->a) a  -- add, sub, zero
-
-def (+)  (d:Add a) ?=> : a -> a -> a = case d of MkAdd add _   _    -> add
-def (-)  (d:Add a) ?=> : a -> a -> a = case d of MkAdd _   sub _    -> sub
-def zero (d:Add a) ?=> : a           = case d of MkAdd _   _   zero -> zero
-
-@instance float64Add : Add Float64 = MkAdd (\x:Float64 y:Float64. %fadd x y) (\x y. %fsub x y) (FToF64 0.0)
-@instance float32Add : Add Float32 = MkAdd (\x:Float32 y:Float32. %fadd x y) (\x y. %fsub x y) (FToF32 0.0)
-@instance int64Add   : Add Int64   = MkAdd (\x:Int64   y:Int64.   %iadd x y) (\x y. %isub x y) (IToI64 0)
-@instance int32Add   : Add Int32   = MkAdd (\x:Int32   y:Int32.   %iadd x y) (\x y. %isub x y) (IToI32 0)
-@instance int8Add    : Add Int8    = MkAdd (\x:Int8    y:Int8.    %iadd x y) (\x y. %isub x y) (IToI8  0)
-@instance unitAdd    : Add Unit    = MkAdd (\x y. ())                        (\x y. ())        ()
-
-@instance tabAdd : Add a ?=> Add (n=>a) =
-  (MkAdd ( \xs ys. for i. xs.i + ys.i )
-         ( \xs ys. for i. xs.i - ys.i )
-         ( for _. zero ))
-
-data Mul a:Type = MkMul (a->a->a) a  -- multiply, one
-
-def (*) (d:Mul a) ?=> : a -> a -> a = case d of MkMul mul _   -> mul
-def one (d:Mul a) ?=> : a           = case d of MkMul _   one -> one
-
-@instance float64Mul : Mul Float64 = MkMul (\x:Float64 y:Float64. %fmul x y) (FToF64 1.0)
-@instance float32Mul : Mul Float32 = MkMul (\x:Float32 y:Float32. %fmul x y) (FToF32 1.0)
-@instance int64Mul   : Mul Int64   = MkMul (\x:Int64   y:Int64.   %imul x y) (IToI64 1)
-@instance int32Mul   : Mul Int32   = MkMul (\x:Int32   y:Int32.   %imul x y) (IToI32 1)
-@instance int8Mul    : Mul Int8    = MkMul (\x:Int8    y:Int8.    %imul x y) (IToI8  1)
-@instance unitMul    : Mul Unit    = MkMul (\x y. ()) ()
+interface Add a:Type where
+  add : a -> a -> a
+  sub : a -> a -> a
+  zero : a
+
+def (+)  (d:Add a) ?=> : a -> a -> a = add
+def (-)  (d:Add a) ?=> : a -> a -> a = sub
+
+@instance float64Add : Add Float64 where
```

Comment:
Perhaps let's remove the `@` now that this is special cased in the parser? This should work similarly to how `interface` works.